ethereum/contracts/query/QueryResponse.sol

```diff
@@ -0,0 +1,220 @@
+    }
+
+    /// @dev ParsedPerChainQueryResponse describes a single per-chain response.
+    struct ParsedPerChainQueryResponse {
+        uint16 chainId;
+        uint8 queryType;
+        bytes request;
+        bytes response;
+    }
+
+    /// @dev EthCallQueryResponse describes an ETH call per-chain query.
+    struct EthCallQueryResponse {
+        bytes requestBlockId;
+        uint64 blockNum;
+        bytes32 blockHash;
+        uint64 blockTime;
+        EthCallData [] result;
+    }
+
+    /// @dev EthCallData describes a single ETH call query / response pair.
+    struct EthCallData {
+        address contractAddress;
+        bytes callData;
+        bytes result;
+    }    
+
+    bytes public constant responsePrefix = bytes("query_response_0000000000000000000|");
+
+    /// @dev getResponseHash computes the hash of the specified query response.
+    function getResponseHash(bytes memory response) public pure returns (bytes32) {
+        return keccak256(response);
+    }
+
+    /// @dev getResponseDigest computes the digest of the specified query response.
+    function getResponseDigest(bytes memory response) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(responsePrefix,getResponseHash(response)));
+    }
+    
+    /// @dev parseAndVerifyQueryResponse verifies the query response and returns the parsed response.
+    function parseAndVerifyQueryResponse(address wormhole, bytes memory response, IWormhole.Signature[] memory signatures) public view returns (ParsedQueryResponse memory r) {
+        verifyQueryResponseSignatures(wormhole, response, signatures);
+
+        uint index = 0;
+        
+        r.version = response.toUint8(index);
+        require(r.version == 1, "invalid response version");
+        index += 1;
+
+        r.senderChainId = response.toUint16(index);
+        index += 2;
+
+        if (r.senderChainId == 0) {
+            r.requestId = response.slice(index, 65);
+            index += 65;
+        } else {
+            r.requestId = response.slice(index, 32);
+            index += 32;
+        }
+        
+        uint32 len = response.toUint32(index); // query_request_len
+        index += 4;
+        uint reqIdx = index;
+
+        require(response.toUint8(reqIdx) == r.version, "version mismatch between request and response");
+        reqIdx += 1;
+
+        r.nonce = response.toUint32(reqIdx);
+        reqIdx += 4;
+
+        uint8 numPerChainQueries = response.toUint8(reqIdx);
+        reqIdx += 1;
+
+        // The response starts after the request.
+        uint respIdx = index + len;
+
+        require(response.toUint8(respIdx) == numPerChainQueries, "num_per_chain_responses does not match num_per_chain_queries");
+        respIdx += 1;
+
+        r.responses = new ParsedPerChainQueryResponse[](numPerChainQueries);
+
+        // Walk through the requests and responses in lock step.
+        for (uint idx = 0; idx < numPerChainQueries; idx++) {
```

Comment:
Use an unchecked loop since the loop is constrained by `numPerChainQueries`. 

`for (uint idx = 0; idx < numPerChainQueries;) {
     // do your thing here
     unchecked { idx += 1; }
}`